### ISSUE TITLE
Remove all use of pointers and references to empty-enum-typed values in Rust code: `enum $name {}`

### ISSUE THREAD

**Issue by briansmith** [briansmith/ring#416]
The semantics of empty enums in Rust aren't well-defined and some of our misuses (along with many others' uses) of that construct are holding back improving the situation in Rust. See https://internals.rust-lang.org/t/recent-change-to-make-exhaustiveness-and-uninhabited-types-play-nicer-together/4602 for a very long discussion. Let's get rid of our misuses to help the Rust language team out.

```rust
pub enum BN_MONT_CTX {}
```

We'll just finish the current work of replacing `BN_MONT_CTX` with `ring::rsa::bigint::Modulus` and ring::rsa::bigint::Mont`.

```rust
pub enum BIGNUM {}
```

In the not-too-distant future, we'll replace `BIGNUM` in a way similar to how we're replacing `BN_MONT_CTX`. But, in the interim, we can just replace this with an actual transliteration of the C definition, since the C definition is so simple:
```
// Keep this in sync with `bignum_st` in include/openssl/bn.h.
mod bignum {
   #[repr(c)]
    pub struct BIGNUM {
        d: *mut Limb,
        top: c::int,
        dmax: c::int,
        neg: c::int,
        flags: c::int,
    }
}
type BIGNUM = bignum::BIGNUM;
```

```rust
pub enum N {}
enum P {}
enum QQ {}
enum Q {}
```

These are using the construct in a reasonable way completely in line with the concept of uninhabited types, since these are only used as `PhantomData` parameters; i.e. we only refer to them at type level, and never refer to values (or references to values, or pointers). So, keep these as-is.

**Comment by briansmith** [comment]
dd27044b8b360440d96f06556af23fad10b2ab98 added one in ring::rand: `enum SecRandomRef {}`. Unfortunately, this is a true opaque type. It is used in a very limited scope, we can use pretty much any alternative hack until https://github.com/rust-lang/rfcs/pull/1861 or something else provides something better.

**Comment by briansmith** [comment]
The `SecRandomRef` part landed in cca54a59c813dce60dcec040838b712da14a2fc7.

**Comment by briansmith** [comment]
`BIGNUM` was converted to regular `#[repr(C)]` structs in 6a030e31caf444fdcc2fcb2738140821c09118ed, as an I suggested above. `BN_MONT_CTX` was also changed the same way in that commit, as an interim change until what I described above can land.
